### PR TITLE
Added batching to prevent the indexer from OOMing MySQL

### LINKED ISSUE #229: Product Indexer causes process to run out of memory
Magento EE 2.2.2
Latest Nosto ver. 2.6.1
Catalog size: 24 000 skus

Full reindex from cli throws this error: 

> SQLSTATE[HY000]: General error: 5 Out of memory (Needed 3140263920 bytes), query was: SELECT COUNT(DISTINCT e.entity_id) FROM `catalog_product_entity` AS `e`
 INNER JOIN `catalog_product_entity_int` AS `at_status_default` ON (`at_status_default`.`row_id` = `e`.`row_id`) AND (`at_status_default`.`attribute_id` = '96') AND `at_status_default`.`store_id` = 0
 LEFT JOIN `catalog_product_entity_int` AS `at_status` ON (`at_status`.`row_id` = `e`.`row_id`) AND (`at_status`.`attribute_id` = '96') AND (`at_status`.`store_id` = 1)
 INNER JOIN `catalog_product_entity_int` AS `at_visibility_default` ON (`at_visibility_default`.`row_id` = `e`.`row_id`) AND (`at_visibility_default`.`attribute_id` = '102') AND `at_visibility_default`.`store_id` = 0 WHERE ((IF(at_status.value_id > 0, at_status.value, at_status_default.value) = 1)) AND (e.created_in <= '1516524480') AND (e.updated_in > '1516524480')

### LINKED ISSUE #275: Full reindex causes the MySQL to crash
Could you improve your index method (\Nosto\Tagging\Model\Indexer\Product\Indexer::executeFull)? It's really don't work on a large collection, for example, we have 160k products - while execution from CLI we get a crash of mysql. But if we using generators or just foreach per batch (in 10k products) - it's ok.

Another issue -

[2018-06-05 10:37:44] nosto.ERROR: Failed to send 50 products for store XXX (2) Error was Invalid product update payload [] []
[2018-06-05 10:37:44] nosto.ERROR: Nosto\Request\Api\Exception\ApiResponseException: Invalid product update payload in 

For ref, magento-e72f949a

### PR DESCRIPTION
**Pull request by mridang**
Closes #275 
Closes #229

### CODE REVIEW

File: Model/Indexer/Product/Indexer.php [hunk 1]
```diff
@@ -57,14 +57,21 @@ public function __construct(
...
+                $pageNumber++;
+
+                $searchCriteria = $this->searchCriteriaBuilder
+                    ->addFilter('status', Status::STATUS_ENABLED, 'eq')
+                    ->setPageSize(self::BATCH_SIZE)
+                    ->setCurrentPage($pageNumber)
+                    ->create();
+                $products = $this->productRepository->getList($searchCriteria);
+                $this->productService->update($products->getItems());
+                $totalItems = $products->getTotalCount();
```
Comment (chenliangde) [inline Model/Indexer/Product/Indexer.php]: put it out of the loop

File: Model/Indexer/Product/Indexer.php [hunk 2]
```diff
@@ -57,14 +57,21 @@ public function __construct(
...
+            $totalItems = 0;
+            do {
+                $pageNumber++;
+
+                $searchCriteria = $this->searchCriteriaBuilder
+                    ->addFilter('status', Status::STATUS_ENABLED, 'eq')
+                    ->setPageSize(self::BATCH_SIZE)
+                    ->setCurrentPage($pageNumber)
+                    ->create();
+                $products = $this->productRepository->getList($searchCriteria);
```
Comment (chenliangde) [inline Model/Indexer/Product/Indexer.php]: check the return value before sending it to the update method

File: Model/Indexer/Product/Indexer.php (hunk 1 above)
Comment (chenliangde) [inline Model/Indexer/Product/Indexer.php]: getTotalCount may return the total number  of products in database. (not the number of products fetched this time). I am not 100% sure, we should at least check it

File: Model/Indexer/Product/Indexer.php [hunk 3]
```diff
@@ -21,7 +21,7 @@
  */
 class Indexer implements IndexerActionInterface, MviewActionInterface
 {
-    const HARD_LIMIT_FOR_PRODUCTS = 10000000;
+    const BATCH_SIZE = 1000;
```
Comment (chenliangde) [inline Model/Indexer/Product/Indexer.php]: How about 200

Comment (supercid) [reply Model/Indexer/Product/Indexer.php]: Why? I think that if the batch size is too small, we'll fall into too many connections issue with MySQL

Comment (chenliangde) [reply Model/Indexer/Product/Indexer.php]: You are right. Keep it.

File: Model/Indexer/Product/Indexer.php (hunk 1 above)
Comment (mridang) [reply Model/Indexer/Product/Indexer.php]: Where? 🤔

File: Model/Indexer/Product/Indexer.php (hunk 3 above)
Comment (mridang) [reply Model/Indexer/Product/Indexer.php]: Magento uses 1000 internally.

File: Model/Indexer/Product/Indexer.php (hunk 2 above)
Comment (mridang) [reply Model/Indexer/Product/Indexer.php]: Was it being checked before?

Comment (chenliangde) [reply Model/Indexer/Product/Indexer.php]: no, but it could return empty now. For say the $totalItems % 1000 == 0, it could case issue.

**Pull request merged**